### ISSUE TITLE
Cannot copy param 0 weights from layer 'fc6'; shape mismatch.

### ISSUE THREAD

**Issue by jimpoon** [shenwei1231/caffe-DeepRegressionForests#2]
HI,

I got an error when trying to run the classify.py included in the caffe-drf/python. How to fix it?

command:
export PYTHONPATH=caffe-drf/python
python classify.py --model_def tmp/Morph-test.prototxt --pretrained_model model/Morph_iter_9000.caffemodel  test.jpg  out.jpg

Cannot copy param 0 weights from layer 'fc6'; shape mismatch.  Source param shape is 4096 25088 (102760448); target param shape is 4096 32768 (134217728).

**Comment by shenwei1231** [comment]
Hi,
   Please follow our instruction to run our code. Why do you use classify.py? This is a script for image classification, which may include an image cropping operation.
Best regards,



Dr. Wei Shen

Assistant Research Professor
Department of Computer Science, Johns Hopkins University, Baltimore, MD USA

Email: wshen10@jhu.edu, shenwei1231@gmail.com
Homepage: https://www.cs.jhu.edu/faculty/wei-shen/

From: jimpoon
Date: 2019-05-28 08:55
To: shenwei1231/caffe-DeepRegressionForests
CC: Subscribed
Subject: [shenwei1231/caffe-DeepRegressionForests] Cannot copy param 0 weights from layer 'fc6'; shape mismatch. (#2)
HI,
I got an error when trying to run the classify.py included in the caffe-drf/python. How to fix it?
command:
export PYTHONPATH=caffe-drf/python
python classify.py --model_def tmp/Morph-test.prototxt --pretrained_model model/Morph_iter_9000.caffemodel test.jpg out.jpg
Cannot copy param 0 weights from layer 'fc6'; shape mismatch. Source param shape is 4096 25088 (102760448); target param shape is 4096 32768 (134217728).
—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or mute the thread.
--

**Comment by jimpoon** [comment]
i see. 

Do you have an reference python script to load and run our trained Deep Regression Forests model file against a face image?

**Comment by jimpoon** [comment]
for your reference, I am using the following code but still got an error while loading the trained model:
import sys
sys.path.insert(0, 'caffe-drf/python')
import caffe

model = 'tmp/Morph-deploy.prototxt';
weights = 'model/Morph_iter_11000.caffemodel';

caffe.set_mode_gpu();
caffe.set_device(0);

net = caffe.Net(model, weights, caffe.TEST);



......
I0528 23:47:35.544059  9673 net.cpp:228] conv1_1 does not need backward computation.
I0528 23:47:35.544062  9673 net.cpp:228] data does not need backward computation.
I0528 23:47:35.544065  9673 net.cpp:270] This network produces output pred
I0528 23:47:35.544080  9673 net.cpp:283] Network initialization done.
F0528 23:47:35.949221  9673 net.cpp:774] Cannot copy param 0 weights from layer 'fc6'; shape mismatch.  Source param shape is 4096 25088 (102760448); target param shape is 4096 32768 (134217728). To learn this layer's parameters from scratch rather than copying from a saved net, rename the layer.
*** Check failure stack trace: ***
Aborted (core dumped)

**Comment by shenwei1231** [comment]
Please check the input dimension indicated in your deploy file, which is not the same as the one defined in the model.




Wei Shen


From: jimpoon
Date: 2019-05-28 11:49
To: shenwei1231/caffe-DeepRegressionForests
CC: Wei Shen; Comment
Subject: Re: [shenwei1231/caffe-DeepRegressionForests] Cannot copy param 0 weights from layer 'fc6'; shape mismatch. (#2)
for your reference, I am using the following code but still got an error while loading the trained model:
import sys
sys.path.insert(0, 'caffe-drf/python')
import caffe
model = 'tmp/Morph-deploy.prototxt';
weights = 'model/Morph_iter_11000.caffemodel';
caffe.set_mode_gpu();
caffe.set_device(0);
net = caffe.Net(model, weights, caffe.TEST);
......
I0528 23:47:35.544059 9673 net.cpp:228] conv1_1 does not need backward computation.
I0528 23:47:35.544062 9673 net.cpp:228] data does not need backward computation.
I0528 23:47:35.544065 9673 net.cpp:270] This network produces output pred
I0528 23:47:35.544080 9673 net.cpp:283] Network initialization done.
F0528 23:47:35.949221 9673 net.cpp:774] Cannot copy param 0 weights from layer 'fc6'; shape mismatch. Source param shape is 4096 25088 (102760448); target param shape is 4096 32768 (134217728). To learn this layer's parameters from scratch rather than copying from a saved net, rename the layer.
*** Check failure stack trace: ***
Aborted (core dumped)

**Comment by jimpoon** [comment]
Hi,

I have checked my deploy prototxt (which is using the one generated from your run.py). Its input layer is like this （256x256)

layer {
  name: "data"
  type: "Input"
  top: "data"
  input_param {
    shape {
      dim: 1
      dim: 3
      dim: 256
      dim: 256
    }
  }
}


Also checked the training prototxt. It seems that they both matched.

layer {
  name: "data"
  type: "ImageData"
  top: "data"
  top: "label"
  transform_param {
    mirror: true
    crop_size: 224
    mean_value: 112.0
  }
  image_data_param {
    source: "./train.txt"
    batch_size: 16
    shuffle: true
    new_height: 256
    new_width: 256
    root_folder: ""
  }
}

I have rewritten my testing python code. It seems that the fc6 weight size mismatch error came out right after the following without any forward inference. 

net = caffe.Net(model, weights, caffe.TEST); 

Any hints about this?

Thx

**Comment by shenwei1231** [comment]
Hi,
   I'm not sure the deploy prototxt you pasted here is correct, as it is exactly the same as the training prototxt.




Wei Shen

From: jimpoon
Date: 2019-05-28 20:21
To: shenwei1231/caffe-DeepRegressionForests
CC: Wei Shen; Comment
Subject: Re: [shenwei1231/caffe-DeepRegressionForests] Cannot copy param 0 weights from layer 'fc6'; shape mismatch. (#2)
Hi,
I have checked my deploy prototxt (which is using the one generated from your run.py). Its input layer is like this （256x256)
layer {
name: "data"
type: "ImageData"
top: "data"
top: "label"
transform_param {
mirror: true
crop_size: 224
mean_value: 112.0
}
image_data_param {
source: "./train.txt"
batch_size: 16
shuffle: true
new_height: 256
new_width: 256
root_folder: ""
}
Also checked the training prototxt. It seems that they both matched.
layer {
name: "data"
type: "ImageData"
top: "data"
top: "label"
transform_param {
mirror: true
crop_size: 224
mean_value: 112.0
}
image_data_param {
source: "./train.txt"
batch_size: 16
shuffle: true
new_height: 256
new_width: 256
root_folder: ""
}
}
I have rewritten my testing python code. It seems that the weight size mismatch error came out right after the following without any forward inference.
net = caffe.Net(model, weights, caffe.TEST);
Any hints about this?
Thx

**Comment by jimpoon** [comment]
Hi,

After changing the tmp/Morph_deploy.prototxt input layer from:
layer {
  name: "data"
  type: "Input"
  top: "data"
  input_param {
    shape {
      dim: 1
      dim: 3
      dim: 256
      dim: 256
    }
  }
}

To:
layer {
  name: "data"
  type: "Input"
  top: "data"
  input_param {
    shape {
      dim: 1
      dim: 3
      dim: 224
      dim: 224
    }
  }
}

everything works fine!

thx.
